# Lecture 4: Summary Statistics and Boulder Weather Data (SOLUTIONS)
***

In this notebook we'll: 
- Compute summary statistics on Boulder weather data 
- Figure out how summary statistics like mean and standard deviation change under transformations of the data
 

The data we'll explore in this notebook concerns temperatures and other weather observations in Boulder County over the month of July 2017.  The data was obtained from the National Oceanic and Atmospheric Administration's [Climate.gov](https://www.climate.gov/) website.  You can find and download loads of climate-related data from NOAA [here](https://www.climate.gov/maps-data/datasets).   

The data is stored in a .csv file called clean_boulder_weather.csv.  

In [2]:
# First we read in the data
# 1) Code Here
weather_data = read.csv('clean_boulder_weather.csv')

In [54]:
# Next we examine the first few rows of the data file using the head() function.
# 2) Code Here
head(weather_data,10)

STATION     NAME                DATE       PRCP TMAX TMIN
1  USW00094075 BOULDER 14 W, CO US 2017-07-01 0.00 68   31  
2  USW00094075 BOULDER 14 W, CO US 2017-07-02 0.00 73   35  
3  USW00094075 BOULDER 14 W, CO US 2017-07-03 0.00 68   46  
4  USW00094075 BOULDER 14 W, CO US 2017-07-04 0.05 68   43  
5  USW00094075 BOULDER 14 W, CO US 2017-07-05 0.01 73   40  
6  USW00094075 BOULDER 14 W, CO US 2017-07-06 0.00 76   48  
7  USW00094075 BOULDER 14 W, CO US 2017-07-07 0.02 74   43  
8  USW00094075 BOULDER 14 W, CO US 2017-07-08 0.00 65   44  
9  USW00094075 BOULDER 14 W, CO US 2017-07-09 0.01 73   39  
10 USW00094075 BOULDER 14 W, CO US 2017-07-10 0.01 75   44

Investigate what the very useful `summary()` function does.

In [53]:
# 3) Code Here
summary(weather_data)

   STATION              NAME               DATE                PRCP        
 Length:214         Length:214         Length:214         Min.   :0.00000  
 Class :character   Class :character   Class :character   1st Qu.:0.00000  
 Mode  :character   Mode  :character   Mode  :character   Median :0.00000  
                                                          Mean   :0.04346  
                                                          3rd Qu.:0.02750  
                                                          Max.   :0.69000  
                                                          NA's   :32       
      TMAX             TMIN      
 Min.   : 54.00   Min.   :31.00  
 1st Qu.: 74.00   1st Qu.:47.00  
 Median : 83.00   Median :55.00  
 Mean   : 81.48   Mean   :53.59  
 3rd Qu.: 89.00   3rd Qu.:60.75  
 Max.   :101.00   Max.   :68.00  
                                 

From this you should see that each row in the DataFrame refers to a particular weather station / date combination.  The columns of the DataFrame are as follows: 

- **STATION**: The unique identification code for each weather station 
- **NAME**: The location / name of the weather station 
- **DATE**: The date of the observation 
- **PRCP**: The precipitation (in inches)
- **TMAX**: The daily maximum temperature (in Fahrenheit)
- **TMIN**: The daily minimum temperature (in Fahrenheit)

From the printed DataFrame above you can see that we actually have data from multiple weather stations.  To see how many, we can pass the **NAME** column (or the **STATION** column) into R's unique function.

In [7]:
# 4) Code Here
unique(weather_data["NAME"])

NAME                     
1   BOULDER 14 W, CO US      
32  GROSS RESERVOIR, CO US   
63  SUGARLOAF COLORADO, CO US
94  NIWOT, CO US             
125 BOULDER, CO US           
155 RALSTON RESERVOIR, CO US 
185 NORTHGLENN, CO US

It looks like we have data from seven different weather stations.  For consistency, let's reduce the data to just the reports from the weather station in Niwot.  

### Exercise 1
***
Extract the rows of the DataFrame concerned with the Niwot weather station.  Store this data in a new DataFrame called dfNiwot.

In [9]:
# 5) Code Here
dfNiwot = weather_data[weather_data$NAME == "NIWOT, CO US",]

In [10]:
dfNiwot

STATION     NAME         DATE       PRCP TMAX TMIN
94  USS0005J42S NIWOT, CO US 2017-07-01 0.0  69   32  
95  USS0005J42S NIWOT, CO US 2017-07-02 0.0  73   37  
96  USS0005J42S NIWOT, CO US 2017-07-03 0.0  68   47  
97  USS0005J42S NIWOT, CO US 2017-07-04 0.1  70   41  
98  USS0005J42S NIWOT, CO US 2017-07-05 0.0  74   40  
99  USS0005J42S NIWOT, CO US 2017-07-06 0.0  78   47  
100 USS0005J42S NIWOT, CO US 2017-07-07 0.0  74   44  
101 USS0005J42S NIWOT, CO US 2017-07-08 0.0  67   43  
102 USS0005J42S NIWOT, CO US 2017-07-09 0.0  75   41  
103 USS0005J42S NIWOT, CO US 2017-07-10 0.0  75   44  
104 USS0005J42S NIWOT, CO US 2017-07-11 0.0  75   52  
105 USS0005J42S NIWOT, CO US 2017-07-12 0.1  54   43  
106 USS0005J42S NIWOT, CO US 2017-07-13 0.0  69   41  
107 USS0005J42S NIWOT, CO US 2017-07-14 0.0  69   40  
108 USS0005J42S NIWOT, CO US 2017-07-15 0.0  68   42  
109 USS0005J42S NIWOT, CO US 2017-07-16 0.0  71   39  
110 USS0005J42S NIWOT, CO US 2017-07-17 0.0  74   47  
111 USS0005J42S NIWOT, CO US 2017-07-18 0.0  71   46  
112 USS0005J42S NIWOT, CO US 2017-07-19 0.0  80   46  
113 USS0005J42S NIWOT, CO US 2017-07-20 0.0  75   47  
114 USS0005J42S NIWOT, CO US 2017-07-21 0.2  74   48  
115 USS0005J42S NIWOT, CO US 2017-07-22 0.2  66   44  
116 USS0005J42S NIWOT, CO US 2017-07-23 0.2  66   41  
117 USS0005J42S NIWOT, CO US 2017-07-24 0.1  74   47  
118 USS0005J42S NIWOT, CO US 2017-07-25 0.2  70   47  
119 USS0005J42S NIWOT, CO US 2017-07-26 0.3  64   46  
120 USS0005J42S NIWOT, CO US 2017-07-27 0.0  65   44  
121 USS0005J42S NIWOT, CO US 2017-07-28 0.1  71   46  
122 USS0005J42S NIWOT, CO US 2017-07-29 0.2  58   48  
123 USS0005J42S NIWOT, CO US 2017-07-30 0.2  62   41  
124 USS0005J42S NIWOT, CO US 2017-07-31 0.0  66   39

### Exercise 2  
***

R has canned functions that compute each of the summary statistics discussed in lecture.  We'll use the mean( ) function as an example.  

Using the mean() function, find the sample mean of the maximum daily temperature in Niwot. 

In [12]:
# 6) Code Here
mean(dfNiwot$TMAX)

[1] 69.83871

Let's see what happens if we call mean( ) on the entire DataFrame. 

In [17]:
# 7) Code Here
mean(dfNiwot)

Warning message in mean.default(dfNiwot):
“argument is not numeric or logical: returning NA”


[1] NA

In [19]:
library(dplyr)
dfNiwot %>%
summarise_if(is.numeric, mean, na.rm=T)


Attaching package: ‘dplyr’


The following objects are masked from ‘package:stats’:

    filter, lag


The following objects are masked from ‘package:base’:

    intersect, setdiff, setequal, union




PRCP       TMAX     TMIN    
1 0.06129032 69.83871 43.54839

The functions for the other summary statistics are as follows: 

\begin{array}{l|l}
\textrm{Function} & \textrm{Statistics} \\
\hline
\textrm{var()} & \textrm{variance} \\
\textrm{sd()} & \textrm{standard deviation} \\
\textrm{min()} & \textrm{minimum value} \\
\textrm{max()} & \textrm{maximum value} \\
\textrm{median()} & \textrm{value} \\
\textrm{quantile(data, probs=c(...)} & \textrm{quantile, where data is the desired input and probs specifies the desired percentile(s) as a decimal} \\
\end{array}

Your job is to use these functions to compute the 5-number summary for the maximum daily temperature for the Niwot weather station. 

In [26]:
# 8) Code Here
minval = min(dfNiwot$TMAX) 
maxval = max(dfNiwot$TMAX) 
Q1 = quantile(dfNiwot$TMAX,.25) 
Q2 = quantile(dfNiwot$TMAX,.50) 
Q3 = quantile(dfNiwot$TMAX,.75) 
cat("5-Number Summary: ", minval, Q1, Q2, Q3, maxval)

5-Number Summary:  54 66.5 70 74 80

### Exercise 3 
***
It turns out that R has a nice function called fivenum( ) that will compute all of the standard summary statistics for you.  

Run the fivenum( ) function on the **TMAX** column of your DataFrame, and check that the results agree with your computations from Exercise 2. 

In [27]:
# 9) Code Here
fivenum(dfNiwot$TMAX)

[1] 54.0 66.5 70.0 74.0 80.0

### Exercise 4 
***
In this exercise we'll explore how the mean and the standard deviation change when we perform basic transformations on the data.  In particular, we're interested in what happens if we 

1. Add or subtract some value from every entry in the data set 
1. Multiply every entry in the data set by some value 

We know from above that the mean and standard deviation of the Niwot **TMAX** value are 69.83871 and 5.621962.  Experiment by adding and multiplying nice integer values with the **TMAX** column and then recomputing the statistics.  From your observations, can you guess how the mean and std dev change under these transformations? 

To get you started, note that the following code will add 3 to each value of the **TMAX** DataFrame:

In [29]:
# 10) Code Here
dfNiwot$TMAX

[1] 69 73 68 70 74 78 74 67 75 75 75 54 69 69 68 71 74 71 80 75 74 66 66 74 70
[26] 64 65 71 58 62 66

In [28]:
dfNiwot$TMAX+3

[1] 72 76 71 73 77 81 77 70 78 78 78 57 72 72 71 74 77 74 83 78 77 69 69 77 73
[26] 67 68 74 61 65 69

In [32]:
print(mean(dfNiwot$TMAX+3))
print(sd(dfNiwot$TMAX+3))

print(mean(dfNiwot$TMAX*3))
print(sd(dfNiwot$TMAX*3))

[1] 72.83871
[1] 5.621962
[1] 209.5161
[1] 16.86588


See if you can prove that your guess works in general mathematically using the formulas for the two statistics: 

$$
\bar{x} = \frac{1}{n} \displaystyle\sum_{k=1}^n x_k \quad \quad \textrm{and} \quad \quad s = \sqrt{\frac{1}{n-1} \sum_{k=1}^n \left( x_k - \bar{x}\right)^2} 
$$

**Solution**: 

**The Mean with Addition**: It looks like when we add a constant to each observation the constant also gets added to the mean.  We can show this in general as follows.  Let $y_k = x_k + a$ be the shifted observations.  We then have  


$$
\bar{y} = \frac{1}{n} \sum_{k=1}^n y_k \quad = \quad \frac{1}{n} \sum_{k=1}^n (x_k + a) \quad = \quad 
\frac{1}{n} \sum_{k=1}^n x_k  + \frac{1}{n} \sum_{k=1}^n a 
\quad = \quad \bar{x}  + \frac{1}{n} \cdot an 
\quad = \quad \bar{x}  + a 
$$

**The Std Dev with Addition**: On the other hand, it looks like the standard deviation stays the same when we add a constant to each observation.  This should make intuitive sense because the std dev is a measure of the spread of the data, and by adding a constant to each observation we're just shifting things down the number line.  Let's see if we can use the formula for std dev to confirm this mathematically. 

$$
\sqrt{\frac{1}{n-1}\sum_{k=1}^n \left( y_k - \bar{y} \right)^2 } \quad = \quad 
\sqrt{\frac{1}{n-1}\sum_{k=1}^n \left[ (x_k + a) - (\bar{x}+a) \right]^2 } \quad = \quad 
\sqrt{\frac{1}{n-1}\sum_{k=1}^n \left( x_k - \bar{x} \right)^2 } 
$$

And so we see that the standard deviation of both the $x$'s and the  $y$'s are the same. 

**The Mean with Multiplication**: It looks like when we multiply each observation by a constant the mean also gets multiplied by the constant.  We can show this in general as follows.  Let $z_k = b \cdot x_k$ be the multiplied observations.  We then have  


$$
\bar{z} = \frac{1}{n} \sum_{k=1}^n z_k \quad = \quad \frac{1}{n} \sum_{k=1}^n b \cdot x_k  \quad = \quad 
b \cdot \frac{1}{n} \sum_{k=1}^n x_k  
\quad = \quad b\cdot \bar{x}  
$$

**The Std Dev with Addition**: It also looks like the std dev gets multiplied by the constant.  Let's see 

$$
\sqrt{\frac{1}{n-1}\sum_{k=1}^n \left( z_k - \bar{z} \right)^2 } \quad = \quad 
\sqrt{\frac{1}{n-1}\sum_{k=1}^n \left( b\cdot x_k - b\cdot \bar{x} \right)^2 } \quad = \quad 
\sqrt{\frac{1}{n-1}\sum_{k=1}^n b^2 \cdot \left( x_k - \bar{x} \right)^2 } \quad = \quad 
b\cdot \sqrt{\frac{1}{n-1}\sum_{k=1}^n  \left( x_k - \bar{x} \right)^2 } 
$$

Confirmed! 

### Exercise 5 
***
OK, let's apply a common transformation to the **TMAX** and **TMIN** columns by converting the temperatures from Fahrenheit to Celsius.  Remember that the transformation is given by 

$$
\textrm{CELSIUS} = \frac{5}{9} (\textrm{FAHRENHEIT}-32) 
$$

First, use the Fahrenheit data in columns **TMAX** and **TMIN** to create Celsius columns in the Niwot DataFrame called **TMAX_C** and **TMIN_C**.

In [36]:
# 11) Code Here
dfNiwot$TMAX_C = (5/9)*(dfNiwot$TMAX - 32)
dfNiwot$TMIN_C = (5/9)*(dfNiwot$TMIN - 32)

In [37]:
dfNiwot

STATION     NAME         DATE       PRCP TMAX TMIN TMAX_C   TMIN_C   
94  USS0005J42S NIWOT, CO US 2017-07-01 0.0  69   32   20.55556  0.000000
95  USS0005J42S NIWOT, CO US 2017-07-02 0.0  73   37   22.77778  2.777778
96  USS0005J42S NIWOT, CO US 2017-07-03 0.0  68   47   20.00000  8.333333
97  USS0005J42S NIWOT, CO US 2017-07-04 0.1  70   41   21.11111  5.000000
98  USS0005J42S NIWOT, CO US 2017-07-05 0.0  74   40   23.33333  4.444444
99  USS0005J42S NIWOT, CO US 2017-07-06 0.0  78   47   25.55556  8.333333
100 USS0005J42S NIWOT, CO US 2017-07-07 0.0  74   44   23.33333  6.666667
101 USS0005J42S NIWOT, CO US 2017-07-08 0.0  67   43   19.44444  6.111111
102 USS0005J42S NIWOT, CO US 2017-07-09 0.0  75   41   23.88889  5.000000
103 USS0005J42S NIWOT, CO US 2017-07-10 0.0  75   44   23.88889  6.666667
104 USS0005J42S NIWOT, CO US 2017-07-11 0.0  75   52   23.88889 11.111111
105 USS0005J42S NIWOT, CO US 2017-07-12 0.1  54   43   12.22222  6.111111
106 USS0005J42S NIWOT, CO US 2017-07-13 0.0  69   41   20.55556  5.000000
107 USS0005J42S NIWOT, CO US 2017-07-14 0.0  69   40   20.55556  4.444444
108 USS0005J42S NIWOT, CO US 2017-07-15 0.0  68   42   20.00000  5.555556
109 USS0005J42S NIWOT, CO US 2017-07-16 0.0  71   39   21.66667  3.888889
110 USS0005J42S NIWOT, CO US 2017-07-17 0.0  74   47   23.33333  8.333333
111 USS0005J42S NIWOT, CO US 2017-07-18 0.0  71   46   21.66667  7.777778
112 USS0005J42S NIWOT, CO US 2017-07-19 0.0  80   46   26.66667  7.777778
113 USS0005J42S NIWOT, CO US 2017-07-20 0.0  75   47   23.88889  8.333333
114 USS0005J42S NIWOT, CO US 2017-07-21 0.2  74   48   23.33333  8.888889
115 USS0005J42S NIWOT, CO US 2017-07-22 0.2  66   44   18.88889  6.666667
116 USS0005J42S NIWOT, CO US 2017-07-23 0.2  66   41   18.88889  5.000000
117 USS0005J42S NIWOT, CO US 2017-07-24 0.1  74   47   23.33333  8.333333
118 USS0005J42S NIWOT, CO US 2017-07-25 0.2  70   47   21.11111  8.333333
119 USS0005J42S NIWOT, CO US 2017-07-26 0.3  64   46   17.77778  7.777778
120 USS0005J42S NIWOT, CO US 2017-07-27 0.0  65   44   18.33333  6.666667
121 USS0005J42S NIWOT, CO US 2017-07-28 0.1  71   46   21.66667  7.777778
122 USS0005J42S NIWOT, CO US 2017-07-29 0.2  58   48   14.44444  8.888889
123 USS0005J42S NIWOT, CO US 2017-07-30 0.2  62   41   16.66667  5.000000
124 USS0005J42S NIWOT, CO US 2017-07-31 0.0  66   39   18.88889  3.888889

Based on the stuff we proved in **Exercise 4**, what do you expect the mean and the standard deviation of the daily maximum temperature to be in Celsius? 

**Solution**: 

Let $\bar{x}^{min}$ and $\bar{x}^{max}$ represent the mean of **TMAX** and **TMIN** in Fahrenheit, respectively, then we expect the means of the associated values in Celsius to be 

$$
\bar{y}^{min} = \frac{5}{9}\left( \bar{x}^{min} - 32 \right) = \frac{5}{9}\left(43.548 -32 \right) = 6.416
\quad \textrm{and} \quad 
\bar{y}^{max} = \frac{5}{9}\left( \bar{x}^{max} - 32 \right) = \frac{5}{9}\left(69.839 -32 \right) = 21.022
$$

Similarly, we expect the standard deviations to be stretched/shrunk by the multiplication factor $\frac{5}{9}$:

In [43]:
cat('Guesses for means in Celsius:', (mean(dfNiwot$TMAX)-32)*5/9, (mean(dfNiwot$TMIN)-32)*5/9, "\n")
cat('Guesses for SDs in Celsius:', (sd(dfNiwot$TMAX))*5/9, (sd(dfNiwot$TMIN))*5/9)

Guesses for means in Celsius: 21.02151 6.415771 
Guesses for SDs in Celsius: 3.123312 2.235326

Once you've made your guess, see if you're right by applying the mean( ) and sd( ) methods to **TMAX_C** and **TMIN_C**. 

In [45]:
mean(dfNiwot$TMAX_C)
mean(dfNiwot$TMIN_C)
sd(dfNiwot$TMAX_C)
sd(dfNiwot$TMIN_C)

[1] 21.02151

[1] 6.415771

[1] 3.123312

[1] 2.235326

### Exercise 6 
***

Compute the daily temperature range (max minus min) for each row in the Niwot DataFrame and store it in a column called **TDIFF**.  Then answer these questions.  

- What is the mean temperature difference over the month of July? 
- What is the difference between the means of the max and min daily temperatures? 
- Do you see a relationship between these two quantities?  If so, can you prove that it's always the case for mean difference and difference of means? 

In [46]:
# 12) Code Here
dfNiwot$TDIFF = dfNiwot$TMAX - dfNiwot$TMIN

In [47]:
dfNiwot

STATION     NAME         DATE       PRCP TMAX TMIN TMAX_C   TMIN_C    TDIFF
94  USS0005J42S NIWOT, CO US 2017-07-01 0.0  69   32   20.55556  0.000000 37   
95  USS0005J42S NIWOT, CO US 2017-07-02 0.0  73   37   22.77778  2.777778 36   
96  USS0005J42S NIWOT, CO US 2017-07-03 0.0  68   47   20.00000  8.333333 21   
97  USS0005J42S NIWOT, CO US 2017-07-04 0.1  70   41   21.11111  5.000000 29   
98  USS0005J42S NIWOT, CO US 2017-07-05 0.0  74   40   23.33333  4.444444 34   
99  USS0005J42S NIWOT, CO US 2017-07-06 0.0  78   47   25.55556  8.333333 31   
100 USS0005J42S NIWOT, CO US 2017-07-07 0.0  74   44   23.33333  6.666667 30   
101 USS0005J42S NIWOT, CO US 2017-07-08 0.0  67   43   19.44444  6.111111 24   
102 USS0005J42S NIWOT, CO US 2017-07-09 0.0  75   41   23.88889  5.000000 34   
103 USS0005J42S NIWOT, CO US 2017-07-10 0.0  75   44   23.88889  6.666667 31   
104 USS0005J42S NIWOT, CO US 2017-07-11 0.0  75   52   23.88889 11.111111 23   
105 USS0005J42S NIWOT, CO US 2017-07-12 0.1  54   43   12.22222  6.111111 11   
106 USS0005J42S NIWOT, CO US 2017-07-13 0.0  69   41   20.55556  5.000000 28   
107 USS0005J42S NIWOT, CO US 2017-07-14 0.0  69   40   20.55556  4.444444 29   
108 USS0005J42S NIWOT, CO US 2017-07-15 0.0  68   42   20.00000  5.555556 26   
109 USS0005J42S NIWOT, CO US 2017-07-16 0.0  71   39   21.66667  3.888889 32   
110 USS0005J42S NIWOT, CO US 2017-07-17 0.0  74   47   23.33333  8.333333 27   
111 USS0005J42S NIWOT, CO US 2017-07-18 0.0  71   46   21.66667  7.777778 25   
112 USS0005J42S NIWOT, CO US 2017-07-19 0.0  80   46   26.66667  7.777778 34   
113 USS0005J42S NIWOT, CO US 2017-07-20 0.0  75   47   23.88889  8.333333 28   
114 USS0005J42S NIWOT, CO US 2017-07-21 0.2  74   48   23.33333  8.888889 26   
115 USS0005J42S NIWOT, CO US 2017-07-22 0.2  66   44   18.88889  6.666667 22   
116 USS0005J42S NIWOT, CO US 2017-07-23 0.2  66   41   18.88889  5.000000 25   
117 USS0005J42S NIWOT, CO US 2017-07-24 0.1  74   47   23.33333  8.333333 27   
118 USS0005J42S NIWOT, CO US 2017-07-25 0.2  70   47   21.11111  8.333333 23   
119 USS0005J42S NIWOT, CO US 2017-07-26 0.3  64   46   17.77778  7.777778 18   
120 USS0005J42S NIWOT, CO US 2017-07-27 0.0  65   44   18.33333  6.666667 21   
121 USS0005J42S NIWOT, CO US 2017-07-28 0.1  71   46   21.66667  7.777778 25   
122 USS0005J42S NIWOT, CO US 2017-07-29 0.2  58   48   14.44444  8.888889 10   
123 USS0005J42S NIWOT, CO US 2017-07-30 0.2  62   41   16.66667  5.000000 21   
124 USS0005J42S NIWOT, CO US 2017-07-31 0.0  66   39   18.88889  3.888889 27

To calculate the mean temperature difference, we need to get the mean of the **TDIFF** column we just created:

In [50]:
# 13) Code Here
mean_diffs = mean(dfNiwot$TDIFF)
cat("The mean of the differences is ", mean_diffs)

The mean of the differences is  26.29032

The difference between the means of the max and min temperatures is:

In [52]:
diff_means = mean(dfNiwot$TMAX)-mean(dfNiwot$TMIN)
cat("The diff of the means is ", diff_means)

The diff of the means is  26.29032

It looks like they're the same! Can we show mathematically that this is true in general? 

**Solution**: 

Suppose we have $n$ observations of two variables $x_k$ and $y_k$ for $k=1,\ldots,n$, and define $z_k = y_k - x_k$.  We want to show that the 

$$
\bar{z} = \bar{y} - \bar{x}.
$$

We'll do this using the definition of the sample mean.  We have 

\begin{eqnarray}
\bar{z} &=& \frac{1}{n}\sum_{k=1}^n z_k  \\
&=& \frac{1}{n}\sum_{k=1}^n (y_k - x_k)   \\
&=& \frac{1}{n}\sum_{k=1}^n y_k - \frac{1}{n}\sum_{k=1}^n x_k  \\ 
&=& \bar{y} - \bar{x}
\end{eqnarray}

**Note** that this little fact is only relevant when we have the same number of observations for $x$ and $y$. 